packages/flutter/lib/src/widgets/sheet_route.dart

```diff
@@ -0,0 +1,266 @@
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'sheet.dart';
+
+/// Route that defines the transition animation of the previous route when this
+/// one is closing
+mixin RouteWithPreviousTransitionMixin<T> on Route<T> {
+  Widget getPreviousRouteTransition(
+    BuildContext context,
+    Animation<double> secondAnimation,
+    Widget child,
+  );
+}
+
+/// Route that allows the next route to define the animation transition of this route
+/// when the route appears back after the next one is popped
+mixin AllowsRouteWithPreviousTransitionMixin<T> on PageRoute<T> {
+  RouteWithPreviousTransitionMixin? _nextModalRoute;
+
+  @override
+  bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
+    return super.canTransitionTo(nextRoute) ||
+        (nextRoute is RouteWithPreviousTransitionMixin);
+  }
+
+  @override
+  void didChangeNext(Route? nextRoute) {
+    if (nextRoute is RouteWithPreviousTransitionMixin) {
+      _nextModalRoute = nextRoute;
+    }
+
+    super.didChangeNext(nextRoute);
+  }
+
+  @override
+  bool didPop(T result) {
+    _nextModalRoute = null;
+    return super.didPop(result);
+  }
+
+  @override
+  Widget buildTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    if (_nextModalRoute != null) {
+      if (secondaryAnimation.isDismissed) {
+        _nextModalRoute = null;
+      } else {
+        // Avoid default transition theme to animate when a new modal view is pushed
+        final Animation<double> fakeSecondaryAnimation =
+            Tween<double>(begin: 0, end: 0).animate(secondaryAnimation);
+
+        final Widget defaultTransition = super.buildTransitions(
```

Comment:
I see your dilemma (in that we never had a pattern to let the bottom route transition different depending on the top route except CupertinoFullscreenDialogTransition which "cheats" by not transitioning the bottom route at all). Though I think we should solve this problem more thoroughly. I see 2 general approaches. Either we create a general plumbing (maybe somewhere here https://github.com/flutter/flutter/blob/b44696626ce329d899f5c29d7664c35b89caad06/packages/flutter/lib/src/widgets/routes.dart#L802) to always let the top route inject a transition to override the transition of the bottom route or we can create something specific for Cupertino and just let CupertinoRouteTransitionMixin be a mixing on ModalRoute instead of PageRoute, then just let it explicitly know and orchestrate between CupertinoFullscreenDialogTransition, CupertinoPageTransition and a new CupertinoSheetTransition (with a bit more work to build in handles to handle all 3 possible top route types in the bottom transition)

Reply:
+1 to finding a more generic way of doing this. I think that should be based in the widget layer, not just cupertino.

Reply:
I am investigating and `TransitionRoute` uses an approach similar to this when a route returns false from canTransitionTo/canTransitionFrom.

The transition route will use `kAlwaysDismissedAnimation` as secondaryAnimation in this case

https://github.com/flutter/flutter/blob/84f3d28555368a70270e9ac8390a9441df95e752/packages/flutter/lib/src/widgets/routes.dart#L344
https://github.com/flutter/flutter/blob/84f3d28555368a70270e9ac8390a9441df95e752/packages/flutter/lib/src/animation/animations.dart#L74-L79